pinot-core/src/main/java/com/linkedin/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -16,143 +16,602 @@
+        new RealtimeSegmentConverter(_realtimeSegment, tempSegmentFolder.getAbsolutePath(), _schema,
+            _segmentZKMetadata.getTableName(), _segmentZKMetadata.getSegmentName(), _sortedColumn, _invertedIndexColumns);
+
+    logStatistics();
+    segmentLogger.info("Trying to build segment");
+    final long buildStartTime = now();
+    try {
+      converter.build(_segmentVersion);
+    } catch (Exception e) {
+      segmentLogger.error("Could not build segment", e);
+      return false;
+    }
+    final long buildEndTime = now();
+    segmentLogger.info("Successfully built segment in {} ms", (buildEndTime - buildStartTime));
+    File destDir = makeSegmentDirPath();
+    FileUtils.deleteQuietly(destDir);
+    try {
+      FileUtils.moveDirectory(tempSegmentFolder.listFiles()[0], destDir);
+      if (buildTgz) {
+        TarGzCompressionUtils.createTarGzOfDirectory(destDir.getAbsolutePath());
+      }
+    } catch (IOException e) {
+      segmentLogger.error("Exception during move/tar segment", e);
+      FileUtils.deleteQuietly(tempSegmentFolder);
+      return false;
+    }
+    FileUtils.deleteQuietly(tempSegmentFolder);
+    return true;
+  }
+
+  private boolean commitSegment() {
+    // Send segmentCommit() to the controller
+    // if that succeeds, swap in-memory segment with the one built.
+    File destSeg = makeSegmentDirPath();
+    final String segTarFileName = destSeg.getAbsolutePath() + TarGzCompressionUtils.TAR_GZ_FILE_EXTENTION;
+    try {
+      _protocolHandler.segmentCommit(_currentOffset, _segmentNameStr, new File(segTarFileName));
+    } catch (FileNotFoundException e) {
+      segmentLogger.error("Tar file {} not found", segTarFileName, e);
+      return false;
+    }
+    return true;
+  }
+
+  private boolean buildSegmentAndReplace() {
+    boolean success = buildSegment(false);
+    if (!success) {
+      return success;
+    }
+    _realtimeTableDataManager.replaceLLSegment(_segmentZKMetadata.getSegmentName());
+    return false;
+  }
+
+  private void hold() {
+    try {
+      Thread.sleep(SegmentCompletionProtocol.MAX_HOLD_TIME_MS);
+    } catch (InterruptedException e) {
+      segmentLogger.warn("Interrupted while holding");
+    }
+  }
+
+  private SegmentCompletionProtocol.Response postSegmentConsumedMsg() {
+    // Post segmentConsumed to current leader.
+    // Retry maybe once if leader is not found.
+    return _protocolHandler.segmentConsumed(_segmentNameStr, _currentOffset);
+  }
+
+  public void goOnlineFromConsuming(RealtimeSegmentZKMetadata metadata) throws InterruptedException {
+    // Takes 30s max to locate a new controller (more if there are multiple controller failures)
+    // set the timeout to 31s;
+    final long transitionStartTimeMs = now();
+    long timeoutMs =_maxTimeForConsumingToOnlineSec * 1000L;
+    LLCRealtimeSegmentZKMetadata llcMetadata = (LLCRealtimeSegmentZKMetadata)metadata;
+    final long endOffset = llcMetadata.getEndOffset();
+    stop(timeoutMs);
```

Comment:
Ah, `stop(timeout)` does the `Thread.join`, it would be better to return a success/failure flag from `stop()` with the reason than checking the current time to know whether or not this succeeded within the time limit.